gpio_controllers/src/gpio_command_controller.cpp

```diff
@@ -0,0 +1,412 @@
+// Copyright 2022 ICUBE Laboratory, University of Strasbourg
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gpio_controllers/gpio_command_controller.hpp"
+
+#include <algorithm>
+
+#include "controller_interface/helpers.hpp"
+#include "hardware_interface/component_parser.hpp"
+#include "rclcpp/logging.hpp"
+#include "rclcpp/qos.hpp"
+#include "rclcpp/subscription.hpp"
+
+namespace
+{
+template <typename T>
+void print_interface(const rclcpp::Logger & logger, const T & command_interfaces)
+{
+  for (const auto & [interface_name, value] : command_interfaces)
+  {
+    RCLCPP_ERROR(logger, "Got %s", interface_name.c_str());
+  }
+}
+
+std::vector<hardware_interface::ComponentInfo> extract_gpios_from_hardware_info(
+  const std::vector<hardware_interface::HardwareInfo> & hardware_infos)
+{
+  std::vector<hardware_interface::ComponentInfo> result;
+  for (const auto & hardware_info : hardware_infos)
+  {
+    std::copy(
+      hardware_info.gpios.begin(), hardware_info.gpios.end(), std::back_insert_iterator(result));
+  }
+  return result;
+}
+}  // namespace
+
+namespace gpio_controllers
+{
+
+GpioCommandController::GpioCommandController() : controller_interface::ControllerInterface() {}
+
+CallbackReturn GpioCommandController::on_init()
+try
+{
+  param_listener_ = std::make_shared<gpio_command_controller_parameters::ParamListener>(get_node());
+  params_ = param_listener_->get_params();
+  return CallbackReturn::SUCCESS;
+}
+catch (const std::exception & e)
+{
+  fprintf(stderr, "Exception thrown during init stage with message: %s \n", e.what());
+  return CallbackReturn::ERROR;
+}
+
+CallbackReturn GpioCommandController::on_configure(const rclcpp_lifecycle::State &)
+try
+{
+  if (!update_dynamic_map_parameters())
+  {
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  store_command_interface_types();
+  store_state_interface_types();
+
+  if (command_interface_types_.empty() && state_interface_types_.empty())
+  {
+    RCLCPP_ERROR(get_node()->get_logger(), "No command or state interfaces are configured");
+    return CallbackReturn::ERROR;
+  }
```

Comment:
If we are having both empty, we are not publishing all the interfaces from the URDF now right?

Reply:
No, we do. The state_interface_types variable stores data from the configuration file, and if state interfaces are empty, it tries to fetch it from urdf. We'll enter this branch when no state and command interfaces are set in config file, and no state interface has been found in urdf.

Reply:
Understood!